operator/pkg/controller/karmada/validator.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2024 The Karmada Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package karmada
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	operatorv1alpha1 "github.com/karmada-io/karmada/operator/pkg/apis/operator/v1alpha1"
+	"github.com/karmada-io/karmada/operator/pkg/util"
+)
+
+// Validator is used to validate Karmada object
+type Validator func(karmada *operatorv1alpha1.Karmada) error
+
+// externalETCDSecretRefValidator the secret name of external etcd should conform to a defined format
+func externalETCDSecretRefValidator(karmada *operatorv1alpha1.Karmada) error {
+	if karmada.Spec.Components.Etcd != nil && karmada.Spec.Components.Etcd.External != nil {
+		expectedSecretName := util.EtcdCertSecretName(karmada.Name)
+		if karmada.Spec.Components.Etcd.External.SecretRef.Name != expectedSecretName {
+			return fmt.Errorf("secret name for external etcd client must be %s, but got %s",
+				expectedSecretName, karmada.Spec.Components.Etcd.External.SecretRef.Name)
+		}
+	}
+	return nil
+}
```

Comment:
I prefer the pattern we used in Webhook.
https://github.com/karmada-io/karmada/blob/f45aedb7161f0fcf94c60be5f3828a5201a70c28/pkg/util/validation/validation.go#L238-L262

So, we don't need to pass the whole object to each `validator` which avoids a lot of nil assertion.